### ISSUE TITLE
Unauthorized even though curl works

### ISSUE THREAD

**Issue by jules-datascientest** [rmartin16/qbittorrent-api#63]
Following my comment on your PR, here's my current problem

Running
```python
import qbittorrentapi

# instantiate a Client using the appropriate WebUI configuration
qbt_client = qbittorrentapi.Client(host='https://qbittorrent.DOMAIN.TLD', port=443, username='USER', password='PASS')

# the Client will automatically acquire/maintain a logged in state in line with any request.
# therefore, this is not necessary; however, you may want to test the provided login credentials.
try:
    qbt_client.auth_log_in()
except qbittorrentapi.LoginFailed as e:
    print(e)

# display qBittorrent info
print(f'qBittorrent: {qbt_client.app.version}')
print(f'qBittorrent Web API: {qbt_client.app.web_api_version}')
for k,v in qbt_client.app.build_info.items(): print(f'{k}: {v}')

# retrieve and show all torrents
for torrent in qbt_client.torrents_info():
    print(f'{torrent.hash[-6:]}: {torrent.name} ({torrent.state})')
```

gives me a 

```
/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host 'qbittorrent.seluj78.fr'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
Traceback (most recent call last):
  File "/home/seluj78/Pytorr/test_qbittorrent_api.py", line 9, in <module>
    qbt_client.auth_log_in()
  File "/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/qbittorrentapi/auth.py", line 92, in auth_log_in
    self._post(
  File "/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/qbittorrentapi/request.py", line 212, in _post
    return self._request_manager(
  File "/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/qbittorrentapi/request.py", line 258, in _request_manager
    return self._request(**kwargs)
  File "/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/qbittorrentapi/request.py", line 290, in _request
    self.handle_error_responses(params=params, response=response)
  File "/home/seluj78/Pytorr/.venv/lib/python3.8/site-packages/qbittorrentapi/request.py", line 542, in handle_error_responses
    raise Unauthorized401Error(response.text)
qbittorrentapi.exceptions.Unauthorized401Error: Unauthorized

Process finished with exit code 1
```

while running

`curl -v -d "username=USER&password=PASS" -X POST https://qbittorrent.DOMAIN.TLD/api/v2/auth/login`

returns

```
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 62.212.90.183:443...
* TCP_NODELAY set
* Connected to qbittorrent.DOMAIN.TLD (62.212.90.183) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=qbittorrent.DOMAIN.TLD
*  start date: Oct 27 04:40:42 2021 GMT
*  expire date: Jan 25 04:40:41 2022 GMT
*  subjectAltName: host "qbittorrent.DOMAIN.TLD" matched cert's "qbittorrent.DOMAIN.TLD"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x5609d5161e10)
> POST /api/v2/auth/login HTTP/2
> Host: qbittorrent.DOMAIN.TLD
> user-agent: curl/7.68.0
> accept: */*
> content-length: 38
> content-type: application/x-www-form-urlencoded
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
* We are completely uploaded and fine
< HTTP/2 200 
< content-security-policy: default-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; script-src 'self' 'unsafe-inline'; object-src 'none'; form-action 'self'; frame-ancestors 'self';
< content-type: text/plain; charset=UTF-8
< date: Wed, 03 Nov 2021 15:57:29 GMT
< referrer-policy: same-origin
< set-cookie: SID=XXXX; HttpOnly; path=/; SameSite=Strict
< x-content-type-options: nosniff
< x-frame-options: SAMEORIGIN
< x-xss-protection: 1; mode=block
< content-length: 3
< 
* Connection #0 to host qbittorrent.DOMAIN.TLD left intact
```

Any clue ?

**Comment by rmartin16** [comment]
The first interesting thing is that `qbittorrentapi.exceptions.Unauthorized401Error` is being raised. While it may seem like a reasonable response for qBittorrent to return when authorization fails, it instead usually means something is going wrong at the HTTP level. (When the credentials don't work, I'm pretty sure qBittorrent just returns a HTTP 200.....but since an authorization cookie wasn't also returned, you aren't actually logged in.....additionally, if you send an invalid auth cookie, qBittorrent will return a HTTP 403.)

Therefore, the 401 is being returned because qBittorrent doesn't like something about your requests.

To establish a better baseline, it may be useful to tweak the `curl` command to actually see if youre being logged in or not. This'll print the auth cookie from qBittorrent if you successfully log in.....it will print `null` (after the `curl` log info is printed) if log in fails.....or at least it should.
```bash
curl -v -d "username=USER&password=PASS" -X POST https://qbittorrent.DOMAIN.TLD/api/v2/auth/login --silent --output /dev/null --cookie-jar - | grep -F "HttpOnly" | awk '{print $7}'
```
This can at least confirm remote log in is indeed possible outside of the context of a browser...(assuming youve already confirmed you can log in to the WebUI directly).

Once it is known that `curl` is working, we need to understand why qBittorrent is returning the 401. My notes say that qBittorrent will most likely return 401 for (at least perceived) XSS or Host Header issues. The qBittorrent log may offer more explicit insight in to why it's rejecting the request. Ultimately, we'll need to know what's upsetting qBittorrent to understand if a change is necessary to this client, the qBittorrent settings (specifically related to HTTP validation), or perhaps the settings for the proxy you're using.

A lot of words....hope this is useful though. Please let me know how further investigation pans out.

**Comment by Seluj78** [comment]
> This can at least confirm remote log in is indeed possible outside of the context of a browser...(assuming youve already confirmed you can log in to the WebUI directly).

I can login to the webUI without a problem :)

I will test the new curl as soon as I'm home and let you know ! 

To me it seems weird that it's a header/proxy issue because when I launch the curl (as shown in the issue) I do get the cookie back ! Anyway we'll keep looking as soon as I'm home

Thanks !

**Comment by rmartin16** [comment]
> I launch the curl (as shown in the issue) I do get the cookie back

ah yeah...i overlooked the SID was replaced with `XXXX`. that's good then we have that confirmed from your initial investigation.

To help narrow down the reason for the rejection, I'd probably start with these settings in qBittorrent. Disabling may stop the auth problem.
![image](https://user-images.githubusercontent.com/387848/140111296-842d9fc4-a4d0-4976-83d1-18f59d4ede15.png)

One of my ideas is this may stem from this client [setting](https://github.com/rmartin16/qbittorrent-api/blob/master/qbittorrentapi/request.py#L439) `Referer` and `Origin` for the requests....

**Comment by Seluj78** [comment]
I'll try those ASAP :)

But it's weird it would be this and not something else ? Or idk. See, for example, radarr, sonarr, etc... All can connect without a problem to my qbitTorrent instance (all hosted on the same docker compose). But I can't access it externally with your API, but it works with curl so I might be those settings yeah ? We'll know soon enough !

**Comment by Seluj78** [comment]
I've just checked, here are the options on my qbit instance:

<img width="654" alt="Screenshot 2021-11-03 at 18 55 36" src="https://user-images.githubusercontent.com/4641317/140165653-1fa96977-65bd-42a8-966a-a85541581f7e.png">

**Comment by Seluj78** [comment]
![image](https://user-images.githubusercontent.com/4641317/140166614-a185a49f-26b6-480c-a082-fe4838c220bd.png)
It worked disabling all three options like shown
<img width="656" alt="Screenshot 2021-11-03 at 19 01 57" src="https://user-images.githubusercontent.com/4641317/140166718-d27a03c1-2632-46d2-86eb-eb75f5c91c67.png">

Is this a bug on your end then ?

**Comment by rmartin16** [comment]
hmm...probably depends on your position. Disabling the security will obviously stop the security errors....but it clearly doesn't really solve any potentially underlying problems. This client is effectively spoofing `Origin` and `Referer` headers since there aren't any really appropriate values but using the domain of the qBittorrent instance should be fine theoretically....although, Im not an HTTP expert by any means.....so, I'm not entirely sure if qBittorrent is rejecting the requests because of those headers.....or if something like `X-Frame-Options` header may be playing a role.

All that said, I'm not particularly inclined to change the default values for those headers.....since they've been this way from the beginning...unless someone has a particularly convincing motivation to do so.

Perhaps to solve your issues with tripping qBittorrent's security alarms....you'll probably need to figure out which headers ultimately stop the errors. Those values could then be hardcoded with the `EXTRA_HEADERS` parameter in this client or using your reverse proxy. Alternatively, it may be completely valid to turn off some of qBittorrent's protection as is sometimes necessary.

**Comment by Seluj78** [comment]
Alright i'll check all of this out :) I'll let you know what I decide to do, thanks :D

**Comment by Seluj78** [comment]
Also I've just tried and it seems it is the CSRF protection that makes the request fail. All other can be enabled

**Comment by rmartin16** [comment]
hmm...ok. I believe that implies qBittorrent thinks your requests are coming from a domain that is different than the one hosting qBittorrent.

Here is where they do the CSRF check: https://github.com/qbittorrent/qBittorrent/blob/master/src/webui/webapplication.cpp#L646

I see they are allowing empty `Referer` headers....but there are also issues against their repo for this initially being a problem. Perhaps that's why I initially defaulted the `Origin` and `Referer` headers....and perhaps they are no longer needed anymore...

At any rate, whatever header (or other value) qBittorrent is getting upset about will be logged to their application log. I'd check there if you want to keep digging....that said, your ultimate exposure from a CSRF attack with qBittorrent is likely quite low anyway...

**Comment by w1ldg00se** [comment]
OP used a https url and specified port 443, which should definitely be a correct configuration. I tried the same and also got the HTTP 401 Unautorized error. When i removed the manually specified port, i.e. `qbittorrentapi.Client(host='https://qbittorrent.DOMAIN.TLD', username='USER', password='PASS')` the login worked and the demo code showed my qbittorrent version and my torrents.

When the HTTP 401 error occurs, qbittorrent logs the error 
```
WebUI: Origin header & Target origin mismatch! Source IP: '::ffff:172.19.0.3'. Origin header: 'https://qbittorrent.DOMAIN.TLD:443/'. Target origin: 'qbittorrent.DOMAIN.TLD'
```
So the issue is that the port is included in the sent headers.
I quickly modified request.py in line 570 to remove the port:
```python
# default headers to prevent qBittorrent throwing any alarms
self._http_session.headers.update(
    {"Referer": self._API_BASE_URL.replace(":443", ""), "Origin": self._API_BASE_URL.replace(":443", "")} if self._API_BASE_URL else {"Referer": self._API_BASE_URL, "Origin": self._API_BASE_URL}
)
```
And then it works also when specifying port 443: `qbittorrentapi.Client(host='https://qbittorrent.DOMAIN.TLD', port=443, username='USER', password='PASS')`.

Note that i run qbittorrent via the docker container lscr.io/linuxserver/qbittorrent and it's behind the proxy traefik, which makes it available on port 443 and handles the certificate needed for https.

Can someone confirm that removing the port from the `Origin` and `Referer` headers works also for users not behind a proxy?

**Comment by rmartin16** [comment]
Ahhhh....this is great. Thank you @w1ldg00se.

An initial thought is that qBittorrent's CSRF check is too naive since it is simply doing a blind comparison of ports between the target URL and origin URL. An HTTPS URL without a port implies port 443....therefore, it seems reasonable for qBittorrent's check to incorporate this well-established standard.

OTOH, this check is frustrating in general for API use outside of the actual WebUI....after all, these Headers will always be spoofed because this client isn't a browser. Given qBittorrent (at least now) doesn't even perform CSRF checks if the `Origin` and `Referer` headers are absent from requests, I'm particularly inclined to stop populating them in requests from this client.

I'll do some more testing (especially with older qBittorrent versions) and report back additional thoughts.

Thanks again.

**Comment by rmartin16** [comment]
I tracked down qBittorrent's history with CSRF.

qbittorrent/qbittorrent@7c2437e5f27c76933c4e1ec3f4734f6cec6c22d8 - CSRF checks implemented (v3.3.13)
qbittorrent/qbittorrent#6882 - community backlash from the new requirement that `Origin` and `Referer` headers must be populated
qbittorrent/qbittorrent#6887 - The reversion to skip CSRF checks if `Origin` and `Referer` aren't populated (v3.3.14)

In qbittorrent/qbittorrent#6882, the devs for sonarr (and really everyone in that issue) say they are willing to spoof headers if necessary but would rather just see the qBittorrent relax this check. All this implies third-party clients did not and are not sending these headers. That gives me confidence I should do the same for this client. (Additionally, this also clarifies for me that CSRF was only actually a problem with the prior API....not version 2.)

Thank you for reporting and investigation.

**Comment by Seluj78** [comment]
You're welcome, thanks for the fix ! 🎉
